### PR TITLE
Add Fast PWM for Controller Fan 

### PR DESCRIPTION
**Pull request by Sangoku**
Added Fast PWM implementation if there is a controller fan  so he does not produce noise

  ### Benefits

When the FLAG FAST_PWM_FAN is enabled it passes the pwm config onto the CONTROLLER_FAN_PIN which is being used. This way the PWM gets applied to the controller and the Attached mosfet.

For some Fans the Coil Whine is really bad. This solves the issue.

### CODE REVIEW

**Comment by marcio-ao** [comment]
We actually do this same modification to our FW, except that instead of using `#if ENABLED(USE_CONTROLLER_FAN)` we use `#if HAS_CONTROLLER_FAN`:

```cpp
  #if HAS_CONTROLLER_FAN
    SET_OUTPUT(CONTROLLER_FAN_PIN);
    #if ENABLED(FAST_PWM_FAN)
      setPwmFrequency(CONTROLLER_FAN_PIN, FAST_PWM_SCALE); // No prescaling. Pwm frequency = F_CPU/256/8
    #endif
  #endif
```

The other thing we do is we added a FAST_PWM_SCALE Configuration option to the second argument to setPwmFrequency for all fan types, which is the sister to SOFT_PWM_SCALE.

File: Marlin/src/module/temperature.cpp [hunk 1]
```diff
@@ -1176,6 +1176,14 @@ void Temperature::init() {
       setPwmFrequency(FAN2_PIN, 1); // No prescaling. Pwm frequency = F_CPU/256/8
     #endif
   #endif
+  
+  #if ENABLED(USE_CONTROLLER_FAN)
```
Comment (marcio-ao) [inline Marlin/src/module/temperature.cpp:1174]: This should be `#if HAS_CONTROLLER_FAN` to mirror the others.

**Comment by marcio-ao** [comment]
I vote for making this standard in Marlin, and also adding a `FAST_PWM_SCALE` constant to `Configuration.h` to control the second argument of `setPwmFrequency` for all fan types. This would make it so we didn't have to make that change to our own FW with every release.

**Comment by thinkyhead** [comment]
> …and also adding a `FAST_PWM_SCALE` constant…

I see a couple of different terms are used for this parameter in Arduino documentation. Would it be better to refer to it as the "scale" or the "divisor"?

**Comment by thinkyhead** [comment]
I see that the actual scaling value differs depending on which pin is being set, so if we want to allow setting specific scale values (1, 8, 32, 64, 128, 256, 1024) then we need to add a method to the HALs to generalize for all MCUs to derive the mode number (1-5 or 1-7) from the divisor and make `setPwmFrequency` smarter.

For example, for ATmega pins 3 and 11 scaling mode 3 produces a scale of 32, while for pins 5, 6, 9, and 10 scaling mode 3 gives a scale of 64.

And/or we could add a separate `*_PWM_SCALE_MODE` for each individual fan.

File: Marlin/src/module/temperature.cpp (hunk 1 above)
Comment (Sangoku) [reply Marlin/src/module/temperature.cpp:1174]: Then i would have to specifically define HAS_CONTROLLER_FAN  in a separate file. As far i can tell the ENABLED is a other version of it without the hassle to redefine it via a check.

Comment (thinkyhead) [reply Marlin/src/module/temperature.cpp:1174]: It's already defined, but you're right to use `ENABLED(USE_CONTROLLER_FAN)`. The other fans are enabled independent of any configuration options, simply on the basis of the pins being defined. (This is a legacy from Marlin 1.0.x.)

**Comment by marcio-ao** [comment]
> ...if we want to allow setting specific scale values (1, 8, 32, 64, 128, 256, 1024) then we need to add a method to the HALs to generalize for all MCUs to derive the mode number

Sounds like a pretty involved change. To be honest, when I tuned it, I simply changed the value until I got the result I needed. I wasn't aiming for a particular frequency or pre-scaler value -- I suspect most people's use case will be similar.

> And/or we could add a separate *_PWM_SCALE_MODE for each individual fan.

That is a good compromise solution. I'm also okay with no change*. For me to update a single value in our FW's code is not a big deal and it sounds like an enhancement to do that from the config might be more complicated than I suspected.

**EDIT:** *To clarify, I still think we should implement @Sangoku original PR, I'm just thinking that making the scaling factor adjustable, as I had requested, may be a lower priority.

**Pull request merged**